content/blog/kubecon-2019/index.md

```diff
@@ -0,0 +1,271 @@
+    ...
+}
+
+func getConfigurationValue(envVar string) string {
+	value := os.Getenv(envVar)
+	if value == "" {
+		log.Fatalf("%s not set", envVar)
+	}
+	log.Printf("%s: %s", envVar, value)
+	return value
+}
+```
+
+### Infrastructure
+
+There are many pieces of infrastructure to spin up and we will use Pulumi to easily wire it all together. Our architecture will look this:
+
+![demo architecture](demo-arch.jpg)
+
+You can check out the Pulumi code that we use to reproduce both our ephemeral environments as well as production in the [infrastructure/index.ts](https://gitlab.com/rocore/demo-app/blob/master/infrastructure/index.ts) file. The neat thing about using Pulumi to spin up both our GCP and Kubernetes resources is that we can create the GCP resources we need and then directly reference them in our Kubernetes deployment. This ensures we're always configuring our application with the correct GCP resources for that environment.
+
+For example, in our Kubernetes deployment, we set the environment variables by using the topic and bucket variables created just above.
+
+```typescript
+// Create a K8s Deployment for our application.
+const appLabels = { appClass: name };
+const deployment = new k8s.apps.v1.Deployment(name, {
+    metadata: { labels: appLabels },
+    spec: {
+        selector: { matchLabels: appLabels },
+        template: {
+            metadata: { labels: appLabels },
+            spec: {
+                containers: [{
+                    ...
+                    env: [
+                        { name: "TOPIC", value: topic.name }, // referencing topic just created
+                        { name: "BUCKET", value: bucket.name }, // referencing bucket just created
+                        { name: "PROJECT", value: project },
+                        {
+                            name: "GOOGLE_APPLICATION_CREDENTIALS",
+                            value: "/var/secrets/google/key.json"
+                        },
+                    ],
+                    ...
+                }]
+            }
+        }
+    },
+});
+```
+
+### Acceptance Tests
+
+Our acceptance tests are what validate that our service when stood up function as expected. They are run against an ephemeral environment. The tests live in the `acceptance/acceptance_test.go` [file](https://gitlab.com/rocore/demo-app/blob/master/acceptance/acceptance_test.go). You'll notice we're once again using that helper function `getConfigurationValue`. Our acceptance test must also be configured to ensure they're validating against the correct resources for that particular ephemeral environment.
+
+Since our service is only accessible from within the Kubernetes cluster, we will use a Kubernetes Job to run our acceptance tests. This is a good technique to use when your CI is running externally, such as from GitLab, and you do not want to publicly expose your service. Our ephemeral environment plus acceptance test looks like this:
```

Comment:
```suggestion
Since our service is only accessible from within the Kubernetes cluster, we use a Kubernetes Job to run our acceptance tests.  Using a Kubernetes job is a good technique to use when your CI is external, and you do not want to expose your service publicly. Our ephemeral environment plus acceptance test looks like this:
```